lib/blocks.php

```diff
@@ -171,98 +172,107 @@ function gutenberg_render_block( $block ) {
  * @return string          Updated post content.
  */
 function do_blocks( $content ) {
-	global $post;
+	$blocks = gutenberg_parse_blocks( $content );
+	return _recurse_blocks( $blocks, $blocks );
+}
+add_filter( 'the_content', 'do_blocks', 9 ); // BEFORE do_shortcode().
 
-	$rendered_content      = '';
-	$dynamic_block_pattern = get_dynamic_blocks_regex();
+/**
+ * Helper function for do_blocks(), to recurse through the block tree.
+ *
+ * @since 4.1.0
+ * @access private
+ *
+ * @param array $blocks Array of blocks from parse_blocks()
+ * @return string
+ */
+function _recurse_blocks( $blocks, $all_blocks ) {
+	global $post;
 
 	/*
 	 * Back up global post, to restore after render callback.
 	 * Allows callbacks to run new WP_Query instances without breaking the global post.
 	 */
 	$global_post = $post;
 
-	while ( preg_match( $dynamic_block_pattern, $content, $block_match, PREG_OFFSET_CAPTURE ) ) {
-		$opening_tag     = $block_match[0][0];
-		$offset          = $block_match[0][1];
-		$block_name      = $block_match[1][0];
-		$is_self_closing = isset( $block_match[4] );
-
-		// Reset attributes JSON to prevent scope bleed from last iteration.
-		$block_attributes_json = null;
-		if ( isset( $block_match[3] ) ) {
-			$block_attributes_json = $block_match[3][0];
+	$rendered_content = '';
+	$dynamic_blocks   = get_dynamic_block_names();
+
+	foreach ( $blocks as $block ) {
+		$block = (array) $block;
```

Comment:
Could we just do something like this, to allow individual plugins to override as necessary for more custom functionality that Core doesn't support w/r/t content, html, inner blocks, etc?

```suggestion
		$block = (array) $block;

		if ( $pre_rendered_content = apply_filters( 'pre_do_block', null, $block, $all_blocks ) ) {
			$rendered_content .= $pre_rendered_content;
			$post = $global_post;
			continue;
		}
```

or potentially even `apply_filters( 'pre_do_block_' . $block['blockName'], null, $block, $all_blocks )` to narrow the filter instead?

Reply:
This could be taken advantage of via something like this:

https://github.com/Automattic/jetpack/pull/10256/commits/a12356bf7ee36cddb2c51cffb3e51861c09cc180